src/metaswitch/clearwater/cluster_manager/etcd_synchronizer.py

```diff
@@ -0,0 +1,185 @@
+                  state == NORMAL) for state in node_states):
+            # All nodes in JOINING_CHANGED_CONFIGURATION,
+            # NORMAL_CHANGED_CONFIGURATION or NORMAL state.
+            return JOINING_RESYNCING
+        elif all((state == NORMAL or
+                  state == WAITING_TO_LEAVE) for state in node_states):
+            # All nodes in NORMAL or WAITING_TO_LEAVE state.
+            return LEAVE_PENDING
+        elif all((state == LEAVING or
+                  state == LEAVING_ACKNOWLEDGED_CHANGE or
+                  state == NORMAL_ACKNOWLEDGED_CHANGE or
+                  state == NORMAL) for state in node_states):
+            # All nodes in LEAVING, LEAVING_ACKNOWLEDGED_CHANGE
+            # NORMAL_ACKNOWLEDGED_CHANGE or NORMAL state.
+            return STARTED_LEAVING
+        elif (all((state == LEAVING_ACKNOWLEDGED_CHANGE or
+                   state == NORMAL_ACKNOWLEDGED_CHANGE or
+                   state == LEAVING_CHANGED_CONFIGURATION or
+                   state == NORMAL_CHANGED_CONFIGURATION)
+                  for state in node_states) and
+              (LEAVING_CHANGED_CONFIGURATION in node_states or
+               NORMAL_CHANGED_CONFIGURATION in node_states)):
+            # At least one node in either LEAVING_ACKNOWLEDGED_CHANGE or
+            # NORMAL_ACKNOWLEDGED_CHANGE, all other nodes in
+            # LEAVING_CHANGED_CONFIGURATION or NORMAL_CHANGED_CONFIGURATION
+            # state.
+            return LEAVING_CONFIG_CHANGING
+        elif all((state == LEAVING_CHANGED_CONFIGURATION or
+                  state == NORMAL_CHANGED_CONFIGURATION or
+                  state == FINISHED or
+                  state == NORMAL) for state in node_states):
+            # All nodes in LEAVING_CHANGED_CONFIGURATION,
+            # NORMAL_CHANGED_CONFIGURATION, FINISHED or NORMAL state.
+            return LEAVING_RESYNCING
+        elif all((state == NORMAL or
+                  state == FINISHED) for state in node_states):
+            # All nodes in NORMAL or FINISHED state.
+            return FINISHED_LEAVING
+        else:
+            # Cluster in unexpected state.
+            return INVALID_CLUSTER_STATE
+
+    # Returns the local node's state in the cluster, and None if the local node
+    # is not in the cluster.
+    def calculate_local_state(self, cluster_view):
+        return cluster_view.get(self._ip)
+
+    # Read the state of the cluster from etcd. The first time we do this, we get
+    # the current state. On subsequent calls we get the most recent state of the
+    # cluster that we haven't previously seen. This may mean waiting for a
+    # change.
+    def read_from_etcd(self):
+        try:
+            if self._index is None:
+                result = self._client.get(self._key)
+            else:
+                result = self._client.eternal_watch(self._key,
+                                                    index=self._index)
```

Comment:
This can fail if a snapshot has occurred since the last read, in which case we should reset `self._index` to `None` and re-do the read.